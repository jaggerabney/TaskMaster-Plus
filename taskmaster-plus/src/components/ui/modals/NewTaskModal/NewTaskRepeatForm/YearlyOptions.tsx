import {
  YearlyOptionsType,
  daysOfTheWeek,
  monthsOfTheYear,
  daysPerMonth,
  setPosDropdownDict
} from "@/utils/Util";

interface YearlyOptionsPropsType {
  data: YearlyOptionsType;
  onChange: (data: YearlyOptionsType) => void;
}

const YearlyOptions: React.FC<YearlyOptionsPropsType> = ({
  data,
  onChange
}) => {
  const { basis, byMonth, byMonthDay, bySetPos, byDay, bySetMonth } = data;

  function inputChangeHandler(
    event:
      | React.ChangeEvent<HTMLInputElement>
      | React.ChangeEvent<HTMLSelectElement>
  ) {
    const targetComponentId = event.target.id;
    let newBasis: string = basis;
    let newByMonth: number = byMonth;
    let newByMonthDay: number = byMonthDay;
    let newBySetPos: number = bySetPos;
    let newByDay: string = byDay;
    let newBySetMonth: number = bySetMonth;

    switch (targetComponentId) {
      case "by-month-checkbox":
        newBasis = "BYMONTH";
        break;
      case "by-month-dropdown":
        newByMonth = monthsOfTheYear.indexOf(event.target.value) + 1;
        break;
      case "by-setpos-checkbox":
        newBasis = "BYSETPOS";
        break;
      case "by-monthday-input":
        newByMonthDay = Number(event.target.value);
        break;
      case "by-setpos-dropdown":
        newBySetPos = Number(event.target.value);
        break;
      case "by-day-dropdown":
        newByDay = event.target.value.slice(0, 2).toUpperCase();
        break;
      case "by-set-month-dropdown":
        newBySetMonth = monthsOfTheYear.indexOf(event.target.value) + 1;
    }

    onChange({
      basis: newBasis,
      byMonth: newByMonth,
      byMonthDay: newByMonthDay,
      bySetPos: newBySetPos,
      byDay: newByDay,
      bySetMonth: newBySetMonth
    });
  }

  return (
    <ul className="flex flex-col text-xl gap-4">
      <li className="flex flex-row items-center gap-4 ml-8">
        <input
          id="by-month-checkbox"
          type="checkbox"
          onChange={inputChangeHandler}
          checked={basis === "BYMONTH"}
        />
        <label htmlFor="by-month-checkbox">On</label>
        <select
          id="by-month-dropdown"
          value={monthsOfTheYear[byMonth - 1]}
          onChange={inputChangeHandler}
          className="border border-black rounded p-2"
        >
          {monthsOfTheYear.map((month) => (
            <option key={month} value={month}>
              {month}
            </option>
          ))}
        </select>
        <input
          id="by-monthday-input"
          type="number"
          step={1}
          min={1}
          value={byMonthDay}
          onChange={inputChangeHandler}
          max={daysPerMonth.get(monthsOfTheYear[byMonthDay - 1])}
          className="border border-black rounded p-2 w-16"
        />
      </li>
      <li className="flex flex-row items-center gap-4 ml-8">
        <input
          id="by-setpos-checkbox"
          type="checkbox"
          onChange={inputChangeHandler}
          checked={basis === "BYSETPOS"}
        />
        <label htmlFor="by-setpos-checkbox">On the</label>
        <select
          id="by-setpos-dropdown"
          value={bySetPos}
          onChange={inputChangeHandler}
          className="border border-black rounded p-2"
        >
          {Array.from(setPosDropdownDict.keys()).map((key) => (
            <option key={key} value={setPosDropdownDict.get(key)}>
              {key}
            </option>
          ))}
        </select>
        <select
          id="by-day-dropdown"
          value={daysOfTheWeek.find((day) =>
            day.toUpperCase().startsWith(byDay)
          )}
          onChange={inputChangeHandler}
          className="border border-black rounded p-2"
        >
          {daysOfTheWeek.map((day) => (
            <option key={day} value={day}>
              {day}
            </option>
          ))}
        </select>
        <label htmlFor="by-set-month-dropdown">of</label>
        <select
          id="by-set-month-dropdown"
          value={monthsOfTheYear[bySetMonth - 1]}
          onChange={inputChangeHandler}
          className="border border-black rounded p-2"
        >
          {monthsOfTheYear.map((month) => (
            <option key={month} value={month}>
              {month}
            </option>
          ))}
        </select>
      </li>
    </ul>
  );
};

export default YearlyOptions;
