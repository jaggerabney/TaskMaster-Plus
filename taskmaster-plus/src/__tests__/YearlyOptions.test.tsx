import "@testing-library/jest-dom";
import { render, fireEvent, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import YearlyOptions from "@/components/ui/modals/NewTaskModal/NewTaskRepeatForm/YearlyOptions";
import {
  defaultRepeatFormState,
  YearlyOptionsType,
  monthsOfTheYear,
  setPosDropdownDict,
  daysOfTheWeek
} from "@/utils/Util";

describe("Yearly options component", () => {
  const mockOnSubmit = jest.fn();
  const mockYearlyOptions = (
    <YearlyOptions
      data={defaultRepeatFormState.yearly}
      onChange={mockOnSubmit}
    />
  );
  let mockOnSubmitCalls = 0;

  it("should render correctly", () => {
    const yearlyOptions = render(mockYearlyOptions);

    expect(yearlyOptions).toMatchSnapshot();
  });

  // basis checkboxes

  it("allows user to change basis", async () => {
    render(mockYearlyOptions);
    const byMonthCheckbox = document.getElementById(
      "by-month-checkbox"
    ) as HTMLInputElement;
    const bySetPosCheckbox = document.getElementById(
      "by-setpos-checkbox"
    ) as HTMLInputElement;
    let onChangeData: YearlyOptionsType;

    await userEvent.click(bySetPosCheckbox);
    onChangeData = mockOnSubmit.mock.calls[mockOnSubmitCalls++][0];
    expect(onChangeData.basis).toEqual("BYSETPOS");

    await userEvent.click(byMonthCheckbox);
    onChangeData = mockOnSubmit.mock.calls[mockOnSubmitCalls++][0];
    expect(onChangeData.basis).toEqual("BYMONTH");
  });

  it('sets the basis to "On date" by default', () => {
    render(mockYearlyOptions);
    const byMonthCheckbox = document.getElementById(
      "by-month-checkbox"
    ) as HTMLInputElement;

    expect(byMonthCheckbox.checked).toBe(true);
  });

  // BYMONTH dropdown

  it("allows user to change BYMONTH dropdown", async () => {
    render(mockYearlyOptions);
    const byMonthDropdown = document.getElementById(
      "by-month-dropdown"
    ) as HTMLSelectElement;
    const augustOption = byMonthDropdown.children[7] as HTMLOptionElement;

    await userEvent.selectOptions(byMonthDropdown, augustOption);

    const onChangeData: YearlyOptionsType =
      mockOnSubmit.mock.calls[mockOnSubmitCalls++][0];
    expect(onChangeData.byMonth).toEqual(8);
  });

  it("only has five BYSETPOS options: 1st, 2nd, 3rd, 4th, and Last", () => {
    render(mockYearlyOptions);
    const byMonthDropdown = document.getElementById(
      "by-month-dropdown"
    ) as HTMLSelectElement;
    const byMonthDropdownOptions = [];

    for (const child of byMonthDropdown.children) {
      byMonthDropdownOptions.push(child.textContent);
    }

    expect(monthsOfTheYear).toEqual(byMonthDropdownOptions);
  });

  it('displays "January" as the default BYMONTH value (On date)', () => {
    render(mockYearlyOptions);
    const byMonthDropdown = document.getElementById(
      "by-month-dropdown"
    ) as HTMLSelectElement;

    expect(byMonthDropdown.options[byMonthDropdown.selectedIndex].text).toEqual(
      "January"
    );
  });

  // BYMONTHDAY input

  it("allows user to change BYMONTHDAY input", () => {
    render(mockYearlyOptions);
    const byMonthDayInput = document.getElementById(
      "by-monthday-input"
    ) as HTMLInputElement;

    fireEvent.change(byMonthDayInput, { target: { value: 12 } });

    const onChangeData: YearlyOptionsType =
      mockOnSubmit.mock.calls[mockOnSubmitCalls++][0];
    expect(onChangeData.byMonthDay).toBe(12);
  });

  it("sets BYMONTHDAY to 1 by default", () => {
    render(mockYearlyOptions);
    const byMonthDayInput = document.getElementById(
      "by-monthday-input"
    ) as HTMLInputElement;

    expect(byMonthDayInput.value).toEqual("1");
  });

  // BYSETPOS dropdown

  it("allows user to change BYSETPOS dropdown", async () => {
    render(mockYearlyOptions);
    const bySetPosDropdown = document.getElementById(
      "by-setpos-dropdown"
    ) as HTMLSelectElement;

    await userEvent.selectOptions(
      bySetPosDropdown,
      screen.getByRole("option", { name: "2nd" })
    );

    const onChangeData: YearlyOptionsType =
      mockOnSubmit.mock.calls[mockOnSubmitCalls++][0];
    expect(onChangeData.bySetPos).toEqual(2);
  });

  it("only has five BYSETPOS options: 1st, 2nd, 3rd, 4th, and Last", () => {
    render(mockYearlyOptions);
    const bySetPosDropdown = document.getElementById(
      "by-setpos-dropdown"
    ) as HTMLSelectElement;
    const setPosDropdownDictKeys = setPosDropdownDict.keys().toArray();
    const bySetPosDropdownOptions = [];

    for (const child of bySetPosDropdown.children) {
      bySetPosDropdownOptions.push(child.textContent);
    }

    expect(setPosDropdownDictKeys).toEqual(bySetPosDropdownOptions);
  });

  it('displays "1st" as the default BYSETPOS value', () => {
    render(mockYearlyOptions);
    const bySetPosDropdown = document.getElementById(
      "by-setpos-dropdown"
    ) as HTMLSelectElement;

    expect(
      bySetPosDropdown.options[bySetPosDropdown.selectedIndex].text
    ).toEqual("1st");
  });

  // BYDAY dropdown

  it("allows user to change BYDAY dropdown", async () => {
    render(mockYearlyOptions);
    const byDayDropdown = document.getElementById(
      "by-day-dropdown"
    ) as HTMLSelectElement;

    await userEvent.selectOptions(
      byDayDropdown,
      screen.getByRole("option", { name: "Tuesday" })
    );

    const onChangeData: YearlyOptionsType =
      mockOnSubmit.mock.calls[mockOnSubmitCalls++][0];
    expect(onChangeData.byDay).toEqual("TU");
  });

  it("only has seven BYDAY options - one for each day of the week", () => {
    render(mockYearlyOptions);
    const byDayDropdown = document.getElementById(
      "by-day-dropdown"
    ) as HTMLSelectElement;
    const byDayDropdownOptions = [];

    for (const child of byDayDropdown.children) {
      byDayDropdownOptions.push(child.textContent);
    }

    expect(daysOfTheWeek).toEqual(byDayDropdownOptions);
  });

  it('displays "Sunday" as the default BYDAY value', () => {
    render(mockYearlyOptions);
    const byDayDropdown = document.getElementById(
      "by-day-dropdown"
    ) as HTMLSelectElement;

    expect(byDayDropdown.options[byDayDropdown.selectedIndex].text).toEqual(
      "Sunday"
    );
  });

  // BYSETMONTH dropdown

  it("allows user to change BYSETMONTH dropdown", async () => {
    render(mockYearlyOptions);
    const bySetMonthDropdown = document.getElementById(
      "by-set-month-dropdown"
    ) as HTMLSelectElement;
    const aprilOption = bySetMonthDropdown.children[3] as HTMLOptionElement;

    await userEvent.selectOptions(bySetMonthDropdown, aprilOption);

    const onChangeData: YearlyOptionsType =
      mockOnSubmit.mock.calls[mockOnSubmitCalls++][0];
    expect(onChangeData.bySetMonth).toEqual(4);
  });

  it("has twelve BYSETMONTH options - one for each month of the year", () => {
    render(mockYearlyOptions);
    const bySetMonthDropdown = document.getElementById(
      "by-set-month-dropdown"
    ) as HTMLSelectElement;
    const bySetMonthDropdownOptions = [];

    for (const child of bySetMonthDropdown.children) {
      bySetMonthDropdownOptions.push(child.textContent);
    }

    expect(monthsOfTheYear).toEqual(bySetMonthDropdownOptions);
  });

  it('displays "January" as the default BYSETMONTH value (On date)', () => {
    render(mockYearlyOptions);
    const bySetMonthDropdown = document.getElementById(
      "by-month-dropdown"
    ) as HTMLSelectElement;

    expect(
      bySetMonthDropdown.options[bySetMonthDropdown.selectedIndex].text
    ).toEqual("January");
  });
});
