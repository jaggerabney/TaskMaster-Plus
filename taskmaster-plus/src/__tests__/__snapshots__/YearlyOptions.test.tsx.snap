// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Yearly options component should render correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <ul
        class="flex flex-col text-xl gap-4"
        id="yearly-options"
      >
        <li
          class="flex flex-row items-center gap-4 ml-8"
        >
          <input
            checked=""
            id="by-month-checkbox"
            type="checkbox"
          />
          <label
            for="by-month-checkbox"
          >
            On
          </label>
          <select
            class="border border-black rounded p-2"
            id="by-month-dropdown"
          >
            <option
              value="January"
            >
              January
            </option>
            <option
              value="February"
            >
              February
            </option>
            <option
              value="March"
            >
              March
            </option>
            <option
              value="April"
            >
              April
            </option>
            <option
              value="May"
            >
              May
            </option>
            <option
              value="June"
            >
              June
            </option>
            <option
              value="July"
            >
              July
            </option>
            <option
              value="August"
            >
              August
            </option>
            <option
              value="September"
            >
              September
            </option>
            <option
              value="October"
            >
              October
            </option>
            <option
              value="November"
            >
              November
            </option>
            <option
              value="December"
            >
              December
            </option>
          </select>
          <input
            class="border border-black rounded p-2 w-16"
            id="by-monthday-input"
            max="31"
            min="1"
            step="1"
            type="number"
            value="1"
          />
        </li>
        <li
          class="flex flex-row items-center gap-4 ml-8"
        >
          <input
            id="by-setpos-checkbox"
            type="checkbox"
          />
          <label
            for="by-setpos-checkbox"
          >
            On the
          </label>
          <select
            class="border border-black rounded p-2"
            id="by-setpos-dropdown"
          >
            <option
              value="1"
            >
              1st
            </option>
            <option
              value="2"
            >
              2nd
            </option>
            <option
              value="3"
            >
              3rd
            </option>
            <option
              value="4"
            >
              4th
            </option>
            <option
              value="-1"
            >
              Last
            </option>
          </select>
          <select
            class="border border-black rounded p-2"
            id="by-day-dropdown"
          >
            <option
              value="Sunday"
            >
              Sunday
            </option>
            <option
              value="Monday"
            >
              Monday
            </option>
            <option
              value="Tuesday"
            >
              Tuesday
            </option>
            <option
              value="Wednesday"
            >
              Wednesday
            </option>
            <option
              value="Thursday"
            >
              Thursday
            </option>
            <option
              value="Friday"
            >
              Friday
            </option>
            <option
              value="Saturday"
            >
              Saturday
            </option>
          </select>
          <label
            for="by-set-month-dropdown"
          >
            of
          </label>
          <select
            class="border border-black rounded p-2"
            id="by-set-month-dropdown"
          >
            <option
              value="January"
            >
              January
            </option>
            <option
              value="February"
            >
              February
            </option>
            <option
              value="March"
            >
              March
            </option>
            <option
              value="April"
            >
              April
            </option>
            <option
              value="May"
            >
              May
            </option>
            <option
              value="June"
            >
              June
            </option>
            <option
              value="July"
            >
              July
            </option>
            <option
              value="August"
            >
              August
            </option>
            <option
              value="September"
            >
              September
            </option>
            <option
              value="October"
            >
              October
            </option>
            <option
              value="November"
            >
              November
            </option>
            <option
              value="December"
            >
              December
            </option>
          </select>
        </li>
      </ul>
    </div>
  </body>,
  "container": <div>
    <ul
      class="flex flex-col text-xl gap-4"
      id="yearly-options"
    >
      <li
        class="flex flex-row items-center gap-4 ml-8"
      >
        <input
          checked=""
          id="by-month-checkbox"
          type="checkbox"
        />
        <label
          for="by-month-checkbox"
        >
          On
        </label>
        <select
          class="border border-black rounded p-2"
          id="by-month-dropdown"
        >
          <option
            value="January"
          >
            January
          </option>
          <option
            value="February"
          >
            February
          </option>
          <option
            value="March"
          >
            March
          </option>
          <option
            value="April"
          >
            April
          </option>
          <option
            value="May"
          >
            May
          </option>
          <option
            value="June"
          >
            June
          </option>
          <option
            value="July"
          >
            July
          </option>
          <option
            value="August"
          >
            August
          </option>
          <option
            value="September"
          >
            September
          </option>
          <option
            value="October"
          >
            October
          </option>
          <option
            value="November"
          >
            November
          </option>
          <option
            value="December"
          >
            December
          </option>
        </select>
        <input
          class="border border-black rounded p-2 w-16"
          id="by-monthday-input"
          max="31"
          min="1"
          step="1"
          type="number"
          value="1"
        />
      </li>
      <li
        class="flex flex-row items-center gap-4 ml-8"
      >
        <input
          id="by-setpos-checkbox"
          type="checkbox"
        />
        <label
          for="by-setpos-checkbox"
        >
          On the
        </label>
        <select
          class="border border-black rounded p-2"
          id="by-setpos-dropdown"
        >
          <option
            value="1"
          >
            1st
          </option>
          <option
            value="2"
          >
            2nd
          </option>
          <option
            value="3"
          >
            3rd
          </option>
          <option
            value="4"
          >
            4th
          </option>
          <option
            value="-1"
          >
            Last
          </option>
        </select>
        <select
          class="border border-black rounded p-2"
          id="by-day-dropdown"
        >
          <option
            value="Sunday"
          >
            Sunday
          </option>
          <option
            value="Monday"
          >
            Monday
          </option>
          <option
            value="Tuesday"
          >
            Tuesday
          </option>
          <option
            value="Wednesday"
          >
            Wednesday
          </option>
          <option
            value="Thursday"
          >
            Thursday
          </option>
          <option
            value="Friday"
          >
            Friday
          </option>
          <option
            value="Saturday"
          >
            Saturday
          </option>
        </select>
        <label
          for="by-set-month-dropdown"
        >
          of
        </label>
        <select
          class="border border-black rounded p-2"
          id="by-set-month-dropdown"
        >
          <option
            value="January"
          >
            January
          </option>
          <option
            value="February"
          >
            February
          </option>
          <option
            value="March"
          >
            March
          </option>
          <option
            value="April"
          >
            April
          </option>
          <option
            value="May"
          >
            May
          </option>
          <option
            value="June"
          >
            June
          </option>
          <option
            value="July"
          >
            July
          </option>
          <option
            value="August"
          >
            August
          </option>
          <option
            value="September"
          >
            September
          </option>
          <option
            value="October"
          >
            October
          </option>
          <option
            value="November"
          >
            November
          </option>
          <option
            value="December"
          >
            December
          </option>
        </select>
      </li>
    </ul>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
